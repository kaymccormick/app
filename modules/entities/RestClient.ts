import axios from 'axios';
import GetEntitiesResponse from './rest/response/GetEntitiesResponse';
export interface Args {
    baseUri: string;
}

export default class RestClient {
    public baseUri: string;

    public constructor(args: Args) {
        this.baseUri = args.baseUri;
    }
    public getEntities(): Promise<GetEntitiesResponse> {
        return axios.get(`${this.baseUri}/entity`).then(response => {
            if(response.status >= 400) {
                return new Error();
            }
            return response.data;
        }).then(data => {
            const r = new GetEntitiesResponse(data.success, data.result);
            return r;
        });
    }
}
