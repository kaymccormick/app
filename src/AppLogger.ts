import { BasicLogger } from './types';

function _getCallerFile() {
    const originalFunc = Error.prepareStackTrace;

    let callerfile;
    let callerlineno;
    try {
        const err = new Error();

        Error.prepareStackTrace = (myErr, stack) => stack;

// @ts-ignore
const x = err.stack.shift();
        const currentfile = x.getFileName();
//        const currentlineno = x.getLineNumber();
        //      process.stderr.write(`${currentfile} ${currentlineno}\n`);

// @ts-ignore
while (err.stack.length) {
// @ts-ignore
const x2 = err.stack.shift();
            callerfile = x2.getFileName();
            callerlineno = x2.getLineNumber();

            if (currentfile !== callerfile) break;
        }
    } catch (e) {
        console.log(e);
    }

    Error.prepareStackTrace = originalFunc;

    return [callerfile, callerlineno];
}

interface Args {
logger?: BasicLogger;
}
export class AppLogger implements BasicLogger {
private logger?: BasicLogger;
public constructor(args?: Args) {
if(args && args.logger) {
this.logger = args.logger;
}
}
// eslint-disable-next-line @typescript-eslint/no-explicit-any
    public debug(message: string, meta?: any): void {
        let meta2;
        if(meta !== undefined) {
            meta2 = Object.assign({}, meta);
            delete meta2.msg;
        } else {
            meta2 = {};
        }
        let myMsg = message;
        if(meta && meta.msg) {
            myMsg += meta.msg;
        }
	const [ file, line ] = _getCallerFile();
        console.log(`${file}:${line}: ${myMsg}`);
    }
    public error(message: string, meta?: any): void {
        let meta2;
        if(meta !== undefined) {
            meta2 = Object.assign({}, meta);
            delete meta2.msg;
        } else {
            meta2 = {};
        }
        let myMsg = message;
        if(meta && meta.msg) {
            myMsg += meta.msg;
        }
	const [ file, line ] = _getCallerFile();
	              console.log(`${file}:${line}: ERROR ${myMsg}`);
    }

}
