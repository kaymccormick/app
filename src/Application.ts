/** SERVER APPLICATION */
import fs from 'fs';

import path from 'path';
import {Logger} from 'winston';

import { Connection } from 'typeorm';
import {createConnection}from './Factory';
import {ServerConfiguration} from './ServerConfiguration';
import {ServerApplicationModule} from './ServerApplicationModule';

const relModulePath = '../modules';
const moduleRoot = path.join(__dirname, relModulePath);

export class Application {
    /* These are for database entities */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
    public entities: any[] = [];
    public connection?: Connection;
    public get modules(): ServerApplicationModule[] { return this.configuration.modules; }

    public configuration: ServerConfiguration;
    // @ts-ignore
    public logger: Logger;
    public constructor(logger: Logger) {
        this.configuration = new ServerConfiguration();
        this.logger = logger;
        if(logger.debug === undefined) {
            throw new Error('invalid logger');
        }
        this.logger.debug('Application.constructor');
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    public init(): Promise<any> {
        fs.readdirSync(moduleRoot).filter((e): boolean => !/^\./.test(e) && fs.statSync(path.join(moduleRoot, e)).isDirectory()).forEach((ent): void => {
            try {
                // eslint-disable-next-line @typescript-eslint/no-var-requires
                const m = require(path.join(relModulePath, ent + '/ServerModule'));
                const serverModule = new m.ServerModule();
                this.configuration.addModule(serverModule);
            } catch(error) {
            }

        });
        this.entities.length = 0;
        this.modules.forEach((module: ServerApplicationModule): void => {
            this.entities.push(...module.getEntities());
        });

        return createConnection(this.entities).then((connection): boolean => {
            this.connection = connection;
            this.modules.forEach((module: ServerApplicationModule): void => {
                module.setup(this, this.configuration);
            });
            return true;
        });
    }

    public getModule(name: string): ServerApplicationModule|undefined {
        return this.modules.find((m): boolean => m.name === name);
    }
}