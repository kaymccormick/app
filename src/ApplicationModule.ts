import { Configuration } from './Configuration';

export interface Actions {
    [actionName: string]: any;
}

export abstract class ApplicationModule {
    public actions: Actions = {};
    public name: string = '';
    public key: string = '';
    public constructor() {
    }
    public abstract setup(config: Configuration): void
    public abstract getReducers(): any;
    public abstract getInitialState(): any;
}

