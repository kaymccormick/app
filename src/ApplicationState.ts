import { CoreModuleKey } from './types';

/**
 * Applicaton state shape. Would like to deprecate this.
 */

export interface ApplicationState {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    [moduleKey: string]: any;
}
