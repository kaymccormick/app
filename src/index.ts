export { createConnection } from './Factory';
export { Application } from './Application';
