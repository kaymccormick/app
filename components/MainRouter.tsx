import React from 'react';
import { BrowserRouter as Router, Route, Link } from "react-router-dom";
import DefaultView from './DefaultView';
import MainViewLayout, {MainViewLayoutProps} from './MainViewLayout';
import { MainComponentBaseProps } from './types';
import { ApplicationModule } from '../src/ApplicationModule';	
// eslint-disable-next-line @typescript-eslint/no-empty-interface
interface MainRouterProps extends MainComponentBaseProps {
    store: any;
}

export default class MainRouter extends React.Component<MainRouterProps> {
    public render() {
        if(!this.props.site || !this.props.app) {
            return null;
        }
        const app = this.props.app!;
        // @ts-ignore
        console.log(this.props.app.modules);
        // @ts-ignore

        const routes = app.modules!.map((m: ApplicationModule<any>) => <Route key={m.key} path={`/${m.key}`} render={() => React.createElement(React.lazy(() => m.getMainComponent()), { module: m })}/>);
        routes.splice(0, 0, <Route path="/" exact render={() => <DefaultView app={app}/>}/>);
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        return <Router basename="/app"><MainViewLayout app={app} store={this.props.store} site={this.props.site} renderMainView={() => <React.Suspense fallback={<div>loading</div>}>{routes}</React.Suspense>}/></Router>;

    }
}
