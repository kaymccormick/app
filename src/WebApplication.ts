import {Configuration} from './Configuration';
import {ApplicationModule} from './ApplicationModule';
import {Module as ClassModelModule } from '../modules/classModel';
import {Module as EntitiesModule } from '../modules/entities';
import { Store, applyMiddleware, createStore, compose, combineReducers } from 'redux';
import thunk from 'redux-thunk';
import { ApplicationState,Action } from '../model/types';

const modules = [];
const relModulePath = '../modules';

export class WebApplication {
    public store?: Store<ApplicationState, Action>;
    public get modules(): ApplicationModule<any>[] { return this.configuration.modules; }

    public configuration: Configuration;
    public constructor() {
        this.configuration = new Configuration();
    }

    public handleChange() {
    }

    public init(): void {
    /* This is totaly gross! */
        const m = new ClassModelModule();
        m.setup(this, this.configuration);
        this.configuration.addModule(m);
        const m2 = new EntitiesModule();
        m2.setup(this, this.configuration);
        this.configuration.addModule(m2);

        const reducers = this.configuration.collectReducers();
        // @ts-ignore
        this.store = createStore(combineReducers(reducers), compose(applyMiddleware(thunk),  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()));
        this.store!.subscribe(this.handleChange);
    }


}
