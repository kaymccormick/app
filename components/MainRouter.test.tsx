import React from 'react';
import { shallow,mount } from 'enzyme';
import MainRouter from './MainRouter';
import { WebApplication } from '../src/WebApplication';
import RestClient from '@heptet/rest-client';

let restClient= new RestClient({baseUri:'', logDebug: (arg) => console.log(arg)});
let configJs = require('../core.conf').default;
let app = new WebApplication({ config: configJs, restClient });
app.init();

beforeEach(() => {
    configJs = require('../core.conf').default;
    app = new WebApplication({ config: configJs, restClient });
    app.init();
});


test('MainRouter', () => {
    const wrapper = mount(<MainRouter store={app.store!} app={app}/>);
    expect(wrapper).toBeDefined();
    expect(wrapper.html()).toMatchSnapshot();
});
