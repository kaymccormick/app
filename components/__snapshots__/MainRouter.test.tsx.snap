// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainRouter 1`] = `null`;
